src/mlpack/methods/ann/loss_functions/mean_absolute_percentage_error_impl.hpp

```diff
@@ -0,0 +1,73 @@
+/**
+ * @file methods/ann/loss_functions/mean_absolute_percentage_error_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of the Mean Absolute Percentage Error function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_MEAN_ABSOLUTE_PERCENTAGE_ERROR_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_MEAN_ABSOLUTE_PERCENTAGE_ERROR_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "mean_absolute_percentage_error.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+MeanAbsolutePercentageError<InputDataType, OutputDataType>::MeanAbsolutePercentageError(
+  const bool reduction) : reduction(reduction)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+typename InputType::elem_type
+MeanAbsolutePercentageError<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    const TargetType& target)
+{
+  InputType loss = arma::abs((input - target) / target);
+  typename InputType::elem_type lossSum = arma::accu(loss);
+
+  if (reduction)
+    return lossSum;
+    
+  return lossSum / target.n_cols;
+}
```

Comment:
Hey @Aakash-kaushik,
I'm not sure if this is correct either. The term `percentage error` suggests that we should about the loss multiplied by 100 and the term `mean` suggests that we always take mean so we shouldn't ever return sum and always return the mean.
For me, the implementation would be along the following lines :
```
return arma::mean(arma::abs((input - target) / target)) * 100;
```

You could also take a look at Mean Absolute Error loss function and maybe multiplying it by 100 is also correct. Not sure.

If you want some references take a look at [this](https://stats.stackexchange.com/questions/58391/mean-absolute-percentage-error-mape-in-scikit-learn).

Reply:
Hey @kartikdutt18 
Yup it seems Right, i will change it accordingly, Also can you guide me on how to implement the change that @zoq suggested in the above review.

Reply:
I Believe the changes that @zoq suggested involve comparing input and target element wise and then updating output element wise based on the criterion.

Reply:
Sure, There are two ways that you could implement the suggestions. They are mentioned below :
1. Allocate space and loop over the matrix. This should be a simple loop implementing the above function.
2. The second approach uses overloaded comparison operators. If write input > target, I will get a matrix of same size with ones where input(i) > target(i). So you could superimpose the conditions like this : 
ansMatrix = ((input > target) * -2) + 1) / target.
So that way we will get -1 where input > target and 1 elsewhere and then we divide it by target.

Reply:
Also should i Add a infinitesimally small number to the denominator to avoid the divide by zero error ?

Reply:
Sure, makes sense.

Reply:
> Sure, There are two ways that you could implement the suggestions. They are mentioned below :
> 
>     1. Allocate space and loop over the matrix. This should be a simple loop implementing the above function.
> 
>     2. The second approach uses overloaded comparison operators. If we write input > target, we will get a matrix of same size with ones where input(i) > target(i). So you could superimpose the conditions like this :
>        ansMatrix = ((input > target) * -2) + 1) / target.
>        So that way we will get -1 where input > target and 1 elsewhere and then we divide it by target.

I was trying the second solution and it doesn't produces the expected output:
`arma::mat input={3, -0.5, 0, 7};`
`arma::mat target={2.5, 0.2, 2, 8};`
`arma::mat comp = (((input > target)*2)-1)/target;`
`output: 0 0 9223372036854775808 2305843009213693952`
`expected output 0.4000   -5.0000   -0.5000   -0.1250`

Reply:
Try this
```
#include<bits/stdc++.h>
#define ll long long
#define MOD 1000000007
#include <armadillo>
using namespace std;


int main()
{
  ios::sync_with_stdio(false);
  arma::mat A(3, 3, arma::fill::zeros);
  arma::mat B(3, 3, arma::fill::zeros);
  A(0) = 1;
  arma::mat t = (arma::conv_to<arma::mat>::from(A > B) * -2) + 1;
  t.print();
  return 0;
}
```

Reply:
Yup This works, thanks.

Reply:
Also, The derivative would be a bit different since forward method has changed. You would also need to divide by `N` and multiply by 100.

Reply:
Yes already did that locally i will push the changes in 10-15 minutes

Reply:
hi @kartikdutt18 ,
pushed the changes can you check them.